cpp/ql/src/semmle/code/cpp/models/implementations/Strdup.qll

```diff
@@ -43,3 +43,32 @@ class StrdupFunction extends AllocationFunction, ArrayFunction, DataFlowFunction
     output.isReturnValueDeref()
   }
 }
+
+/**
+ * A `strndup` style allocation function.
+ */
+class StrndupFunction extends AllocationFunction, ArrayFunction, TaintFunction {
+  StrndupFunction() {
+    exists(string name |
+      hasGlobalOrStdName(name) and
+      (
+        // strndup(str, maxlen)
+        name = "strndup"
+      )
+    )
+  }
+
+  override predicate hasArrayInput(int bufParam) { bufParam = 0 }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = 0 }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // This function may do only a partial copy of the input buffer to the output
+    // buffer, so it's a taint flow.
+    (
+      input.isParameterDeref(0) or
+      input.isParameter(1)
```

Comment:
I'm not convinced there should be taint from the max-length parameter to the returned buffer. Do you have a use case in mind?

Reply:
I'm currently of the mind that any time A affects the value of B, A should taint B - the only exception being from binary conditions such as in conditional control flow (`if` statements).  In this case, the max-length parameter can cut the string short an arbitrary amount, potentially changing it's meaning (e.g. stripping data from the end of something like a URL).

We had a discussion on Slack on Friday about some similar (ish) situations - string-to-int and int-to-string conversions.  The conclusion seemed to be that taint should flow, and the user can always exclude these kinds of things using the `barrier` predicate.  That said, I can see that in the case of a maximum length parameter the influence is less than an int-to-string conversion.

Whatever we decide, I'd like to make sure we're consistent about this across the models library - and probably even document our choices in `Taint.qll`.

Reply:
Okay, you've convinced me that we should allow this taint flow and that it's different from `memcpy` because we end up with a null termination that actually changes the meaning of the string, which makes it dangerous to use the newly allocated string even if the original string wasn't tainted.